cvmfs/smalloc.h

```diff
@@ -73,6 +73,44 @@ static inline void __attribute__((used)) smunmap(void *mem) {
   assert((retval == 0) && "Invalid umnmap");
 }
 
+
+/**
+ * Used when the caller remembers the size, so that it can call sxunmap later.
+ */
+static inline void * __attribute__((used)) sxmmap(size_t size) {
+  void *mem = mmap(NULL, size, PROT_READ | PROT_WRITE,
+                   MAP_PRIVATE | PLATFORM_MAP_ANONYMOUS, -1, 0);
+  assert((mem != MAP_FAILED) && "Out Of Memory");
+  return mem;
+}
+
+
+/**
+ * Free memory acquired by sxmmap.
+ */
+static inline void __attribute__((used)) sxunmap(void *mem, size_t size) {
+  int retval = munmap(mem, size);
+  assert((retval == 0) && "Invalid umnmap");
+}
+
+
+/**
+ * Pointer is aligned at a multiple of the size.  The size has to be a multiple
+ * of 2MB.
+ */
+static inline void * __attribute__((used)) sxmmap_align(size_t size) {
+  assert((size % (2 * 1024 * 1024)) == 0);
+  char *mem = reinterpret_cast<char *>(sxmmap(2 * size));
+  uintptr_t head = size - (uintptr_t(mem) % size);
+  sxunmap(mem, head);
+  mem += head;
+  uintptr_t tail = size - head;
+  if (tail > 0)
+    sxunmap(mem + size, tail);
+  return mem;
+}
+
+
```

Comment:
Perhaps wrap this `mmap()` into a thin class? Provides you with RAII and keeps the `*mem` and `size` contained for `sxunmap()`.